### DISCUSSION TITLE
DefaultRubyVM(module) crash  using my own wasm file

### DISCUSSION THREAD

**Discussion by jeezs** [ruby/ruby.wasm#259] [Q&A] [answered]
I create a wasm file using the example found in the git : 

`curl -LO https://github.com/ruby/ruby.wasm/releases/latest/download/ruby-3_2-wasm32-unknown-wasi-full.tar.gz
tar xfz ruby-3_2-wasm32-unknown-wasi-full.tar.gz
mv 3_2-wasm32-unknown-wasi-full/usr/local/bin/ruby ruby.wasm
mkdir src
echo "require './src/my_file'\nputs 'Hello'" > src/my_app.rb
echo "puts 'Hello from my REQUIRED file'" > src/my_file.rb
wasi-vfs pack ruby.wasm --mapdir /src::./src --mapdir /usr::./3_2-wasm32-unknown-wasi-full/usr -o my-ruby-app.wasm
wasmtime my-ruby-app.wasm -- /src/my_app.rb`

then I try to execute the wasm file in the browser using the following code : 

\<html>

\<script src="https://cdn.jsdelivr.net/npm/@ruby/wasm-wasi@latest/dist/browser.umd.js"></script>

\<script>
    const {DefaultRubyVM} = window["ruby-wasm-wasi"];
    const main = async () => {
        const response = await fetch(
            "my-ruby-app.wasm"
        );

    const buffer = await response.arrayBuffer();
        const module = await WebAssembly.compile(buffer);
        const {vm} = await DefaultRubyVM(module);
        vm.printVersion();
        vm.eval(`
        require "js"
        luckiness = ["Lucky", "Unlucky"].sample
        JS::eval("document.body.innerText = '#{luckiness}'")
      `);
 
    };
    main();
\</script>
\<body>\</body>
\</html>

The lin e : `const {vm} = await DefaultRubyVM(module);` raise the error `Unhandled Promise Rejection: TypeError: First argument to FinalizationRegistry should be a function`
Maybe I use the wasm file the wrong way or is it a bug or Wasmtime and my wasm file have version compatibility problem . If someone have an idea? or is there another way to use my own wasm file within the browser?

**Comment by jeezs** [comment]
Answering myself:
I was using the wrong install because I thought WASI is the same for JS and native, should install: `ruby-3_2-wasm32-unknown-wasi-full-js.tar.gz` instead of : `ruby-3_2-wasm32-unknown-wasi-full.tar.gz`
The correct install script is :
`curl -LO https://github.com/ruby/ruby.wasm/releases/latest/download/ruby-3_2-wasm32-unknown-wasi-full-js.tar.gz
tar xfz ruby-3_2-wasm32-unknown-wasi-full-js.tar.gz

mv 3_2-wasm32-unknown-wasi-full-js/usr/local/bin/ruby ruby_js.wasm

mkdir src
echo "require './src/my_file'\nputs 'Hello from main file'" > src/my_app.rb
echo "puts 'Hello from my required file'" > src/my_file.rb

wasi-vfs pack ruby_js.wasm --mapdir /src::./src --mapdir /usr::./3_2-wasm32-unknown-wasi-full-js/usr -o my_js_app.wasm`

**Comment by jeezs** [comment]
Now the problem is wasi-vfs that have problem with js profile. Hope it'll be solved soon.

**Comment by kateinoigakukun** [comment]
@jeezs Thank you for detailed report :pray: Which version of wasi-vfs are you using?

**Comment by jeezs** [comment, marked as answer]
the version that causes  trouble is wasi-vfs version 0.4.0. 
So I switch to version 0.3.0. 
latest version of browser.umd.js also conflict with any versions of wasi_vfs so I download this version  instead   : https://cdn.jsdelivr.net/npm/@ruby/wasm-wasi/dist/browser.umd.js, 
Then it works.
